### PR TITLE
Ashley - Connecting_MonteCarlo_to_ModernAI

### PR DESCRIPTION
**Pull request by soltaniehha**
Module under folder `Connecting_MonteCarlo_to_ModernAI`

### CODE REVIEW

File: Connecting_MonteCarlo_to_ModernAI/README.md [hunk 1]
```diff
@@ -22,3 +22,4 @@ Suggested time is 2-3 hours per notebook; total time for the module can be betwe
...
+|N2. 2D Ising Model | Introduces students to the derivation for the Ising Model, and walks them through a 2D simulation of a square Ising lattice.  Students are taught to identify the parts of the algorithm which are compute-intensive, and basic complexity and optimization skills for python code. <br><br> Total Time: | DG1. Implement the Metropolis-Hastings algorithm in Python using `numpy`.<br><br>DG2. Implement basic timing methods for understanding code execution. <br><br>DG3. Implement `numba` to improve code timing. | PG1. Understand the origin of the Ising Model and ferromagnetism. <br><br>PG2. Simulate net-magnetization curves as a function of temperature for a 2D system.<br><br> PG3. Connect the phase transition from magnetic disorder to order with the formation of magnetic domans in a thermalized system.|
```
Comment (cnrrobertson) [inline Connecting_MonteCarlo_to_ModernAI/README.md]: Could entries be added for the 3rd and 4th notebooks as well? Or if you are ultimately not including the 4th notebook, I think a couple updates are needed in the readme to clarify.

**Comment by cnrrobertson** [comment]
I know that the main focus of the modules is to complete them in Colab, but it could be nice to have a `requirements.txt` for pip or an `environment.yml` for anaconda with the packages that would be needed for the module.

File: N1_IntroToMonteCarlo.ipynb [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "view-in-github",
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/GDS-Education-Community-of-Practice/DSECOP/blob/daleas_module/N1_IntroToMonteCarlo.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (cnrrobertson) [inline N1_IntroToMonteCarlo.ipynb]: I think this html link needs to be changed to fix the Colab link after renaming the notebook.

File: 03_Predicting_Inverse_Temperature.ipynb [hunk 3]
```diff
@@ -0,0 +1,1377 @@
+{
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "view-in-github",
+        "colab_type": "text"
+      },
+      "source": [
+        "<a href=\"https://colab.research.google.com/github/GDS-Education-Community-of-Practice/DSECOP/blob/daleas_module/03_Predicting_Inverse_Temperature.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (cnrrobertson) [inline 03_Predicting_Inverse_Temperature.ipynb]: I think this html link also needs to be adjusted with the new name

File: N1_IntroToMonteCarlo.ipynb [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+      "metadata": {
+        "id": "O4Wkbgt-dkCC"
+      },
+      "execution_count": 1,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "# Now we run the simulation\n",
```
Comment (cnrrobertson) [inline N1_IntroToMonteCarlo.ipynb]: This cell works great, but I didn't immediately realize that you have to change the current state back to 'E' in the cell two cells above every time you rerun it. Maybe you could redefine current cell to 'E' in this cell?

File: N1_IntroToMonteCarlo.ipynb [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "\n",
+        "def MarkovChainExperiment():\n",
+        "\n",
+        "  # Begin by initializing the \n",
+        "  # current state with the value passed into the function\n",
+        "  current_state = \n",
```
Comment (cnrrobertson) [inline N1_IntroToMonteCarlo.ipynb]: It might be good to also include `final_state=` here. Or better, put them as parameters to the function `MarkovChainExperiment(current_state, final_state)`?

File: N1_IntroToMonteCarlo.ipynb [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+        "for step in range(num_steps):\n",
+        "  p = get_probability()\n",
+        "\n",
+        "  # --- HERE IS THE STATE MACHINE CODE --- #\n",
+        "\n",
+        "  # --> The first logical case is written for you\n",
+        "\n",
+        "  # If the state is currently 'E' \n",
+        "  # and transition probability is less than or equal to 0.3\n",
+        "  if current_state == 'E' and p<=0.3:\n",
```
Comment (cnrrobertson) [inline N1_IntroToMonteCarlo.ipynb]: I think these if statements should maybe have a `continue` in them? As they are currently, the state is changed and then the other if statements are processed before the step is increased, so the results may be artificially short. What do you think?

File: N1_IntroToMonteCarlo.ipynb (hunk 5 above)
Comment (cnrrobertson) [reply N1_IntroToMonteCarlo.ipynb]: I see now that was the problem prompt... Sorry

File: N1_IntroToMonteCarlo.ipynb (hunk 6 above)
Comment (cnrrobertson) [reply N1_IntroToMonteCarlo.ipynb]: Actually, the `continue` command alone wouldn't be enough. You probably need to have one large if-else to check if there is a current state of E or A and then do the rest inside that.

**Comment by cnrrobertson** [comment]
Hey @daleas0120, I've finished my pass through notebook 1. Looks great overall. Very nice intro to Monte carlo stuff and I think the exercises are the perfect difficulty. I tried to leave specific comments on places that might need just a tad bit more explanation or a little adjustment. It took me about an hour. Mostly spent on reading.

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb [hunk 7]
```diff
@@ -0,0 +1,1233 @@
...
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "AvTQs_NkPdgj"
+      },
+      "outputs": [],
+      "source": [
+        "lattice[lattice>0.5]=1 # any values greater than 0.5, I set equal to 1\n",
+        "lattice[lattice !=1]=-1 # anything I don't change before, I make equal to -1"
+      ]
```
Comment (cnrrobertson) [inline Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb:334]: It may be easier (to do and to explain) to just use `np.random.choice([-1,1], size=(10,10))` which just randomly chooses -1 or 1 for all points.

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb [hunk 8]
```diff
@@ -0,0 +1,1233 @@
...
+        "sqrt_N = 25 \n",
+        "\n",
+        "# Create a new lattice\n",
+        "init_lattice = np.random.uniform(size=(sqrt_N,sqrt_N))\n",
+        "\n",
+        "#mask lattice\n",
+        "init_lattice[init_lattice>0.5]=1\n",
+        "init_lattice[init_lattice !=1]=-1\n",
+        "\n",
+        "# A new step here to create non-interacting atoms around the edge\n",
```
Comment (cnrrobertson) [inline Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb]: Maybe you can add some text explanation to this adjustment? You discuss that it is a challenge before, so it might be clarifying for people to see written out exactly what you're doing to account for it.

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb [hunk 9]
```diff
@@ -0,0 +1,1233 @@
...
+      "source": [
+        "# Complete the initialize_lattice function below\n",
+        "\n",
+        "def initialize_lattice(sqrt_n: int):\n",
+        "    \"\"\"    \n",
+        "    Function to initialize lattice.  Adds a border of zeros\n",
+        "    to represent non-interacting atoms and make the neighbor\n",
+        "    calculation easier\n",
+        "\n",
+        "    sqrt_n: The square root of the number of atoms in the lattice\n",
```
Comment (cnrrobertson) [inline Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb:769]: I was a little confused by `sqrt_n` in the parameters and `L` in all the comments. Took me a second to realize they were the same. I'm not sure the best way to clarify..

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb [hunk 10]
```diff
@@ -0,0 +1,1233 @@
...
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### 1b) Perform the following experiment \n",
+        "\n",
+        "By calling the `main_exercise1` function using different arguments:\n",
```
Comment (cnrrobertson) [inline Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb]: `main_exercise1` needs to be renamed to `main_exercise1b` in the discussion text throughout

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb [hunk 11]
```diff
@@ -0,0 +1,1233 @@
...
+        "\n",
+        "1. Re-run the simulation for lattice sizes of $\\sqrt{N}$ = [10, 15, 20, 25, 40].\n",
```
Comment (cnrrobertson) [inline Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb:846]: For my laptop, it took 10 min of running just to get up to sqrt_n=25. Maybe you could drop the 40? Just so students aren't waiting around too long. Especially if they have a couple runs with mistakes.

**Comment by cnrrobertson** [comment]
Hey @daleas0120, I have gone through notebook 2 and it's awesome. I love that you included a bit of code profiling because you can very easily see the big gains just by using `numba` and it is a very useful tool for the students. It took me probably an hour to go through but mostly spent on the reading.

File: 03_Predicting_Inverse_Temperature.ipynb [hunk 12]
```diff
@@ -0,0 +1,1377 @@
...
+        "    calculation easier\n",
+        "\n",
+        "    L: The square root of the number of atoms in the lattice\n",
+        "    returns padded_lattice: A lattice of size (L+2, L+2)\n",
+        "    \"\"\"\n",
+        "\n",
+        "    # initialize the lattice using np.random.random for a lattice of size\n",
+        "    # LxL\n",
+        "\n",
+        "    init_lattice = np.random.random(size=(L,L))\n",
```
Comment (cnrrobertson) [inline 03_Predicting_Inverse_Temperature.ipynb]: I mentioned this in notebook 2 but it may be easier to explain if this function is rewritten along the lines of
```
init_lattice = np.random.choice([-1,1], size=(L,L))
padded_lattice = np.zeros((L+2,L+2))
padded_lattice[1:-1,1:-1] = init_lattice
return padded_lattice
```

**Comment by cnrrobertson** [comment]
Hey @daleas0120, I have gone through notebook 3. It's really awesome. I don't have any specific spots, but I generally really like the high level overviews you give of the methods and the way you let students just try running things and keeping track of the results. I'll be honest, I didn't do the full training for the CNN because it does take a bit! 

My only concern is that the CNN may be a bit overwhelming without better understanding of neural networks and that the long training time will be prohibitive of the students finishing. To address that, maybe an idea could be:

1. Shorten the "Build a CNN model" description to just give the bare bones info on the layers of the network and then point them to an appendix with the full details as you have them written.
2. You are using a very small learning rate in Adam. Is that needed? I tried with a much larger rate (0.05) and the training after 1 epoch gave a test MSE of 0.01. Maybe you could start them with a larger rate and then also include in the optional that they can try with a smaller rate and more epochs if they'd like to get really good results.

File: Connecting_MonteCarlo_to_ModernAI/README.md (hunk 1 above)
Comment (daleas0120) [reply Connecting_MonteCarlo_to_ModernAI/README.md]: Captured.

Comment (daleas0120) [reply Connecting_MonteCarlo_to_ModernAI/README.md]: Thank you for your comment!  I will update the README as notebooks become available.  The README has been updated to reflect the currently completed work.

File: N1_IntroToMonteCarlo.ipynb (hunk 2 above)
Comment (daleas0120) [reply N1_IntroToMonteCarlo.ipynb]: Thank you for your comment.   The notebook links are specific to the branch name.  I will update all of the Google Colab links simultaneously after this branch is merged with the main branch.

File: 03_Predicting_Inverse_Temperature.ipynb (hunk 3 above)
Comment (daleas0120) [reply 03_Predicting_Inverse_Temperature.ipynb]: Thank you for your comment.   The notebook links are specific to the branch name.  I will update all of the Google Colab links simultaneously after this branch is merged with the main branch.

File: N1_IntroToMonteCarlo.ipynb (hunk 4 above)
Comment (daleas0120) [reply N1_IntroToMonteCarlo.ipynb]: Thank you for your comment.  I have moved the cell with the `current_state` variable to be closer to the simulation cell.  I do not want to combine this variable with the simulation code because I want students to understand the difference between initializing the system and simulating the system.  The difficulty of executing two separate cells also motivates the development of the code function in the next exercise, where this variable is combined with the simulation code.

File: N1_IntroToMonteCarlo.ipynb (hunk 6 above)
Comment (daleas0120) [reply N1_IntroToMonteCarlo.ipynb]: Thank you for your comment.  I have updated three of the `if` statements to `elif` to adjust for the "continue" statement you suggested.  Because the final `if` statement occurs at the same level as the state change `if` statements, it covers the case of a current state E or A.

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb (hunk 7 above)
Comment (daleas0120) [reply Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb:334]: Thank you for your comment.  While I agree that this function is a more elegant solution, it is also hyper-specific to the Python programming language and obscures the idea of preparing a system for simulation.  Other coding languages will still require the student to randomly draw from a distribution and then assign a value.  I have updated the plotting code to use this function, but left the explanation as is.

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb (hunk 9 above)
Comment (daleas0120) [reply Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb:769]: Thank you for your comment.  I have added wording explaining what `sqrt_n` is to the markup, and removed references to variable `L`.

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb (hunk 10 above)
Comment (daleas0120) [reply Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb]: Thank you for your comment.  The correction has been captured.

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb (hunk 8 above)
Comment (daleas0120) [reply Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb]: Thank you for your comment.  The in-line code comment has been updated as follows:

```
# A new step here to create non-interacting atoms around the edge by padding
# the array edges with zeroes.  This way, we can iterate over all N atoms in the 
# system without causing an out-of-bounds error
```

File: Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb (hunk 11 above)
Comment (daleas0120) [reply Connecting_MonteCarlo_to_ModernAI/N2_2D_Ising_Model.ipynb:846]: Thank you for your comment.  Since I was able to do a 40x40 system size in about 10 minutes using Google Colab, I will leave this here, but have added the following note for students in the exercise instructions:

>Note: This computation may take some time to complete. Discuss with your instructor how long you should wait for a simulation to run.

I will also update the Instructor solutions that students running this on their personal machines may need accommodations for different run times.

File: 03_Predicting_Inverse_Temperature.ipynb (hunk 12 above)
Comment (daleas0120) [reply 03_Predicting_Inverse_Temperature.ipynb]: Thank you for your comment.  Since I am explicitly reusing the code from N2, I do not want to introduce changes here that are not in N2.  See comment above for why I am not changing the main algorithm in N2 to use this function.

**Comment by daleas0120** [comment]
> Hey @daleas0120, I have gone through notebook 3. It's really awesome. I don't have any specific spots, but I generally really like the high level overviews you give of the methods and the way you let students just try running things and keeping track of the results. I'll be honest, I didn't do the full training for the CNN because it does take a bit!
> 
> My only concern is that the CNN may be a bit overwhelming without better understanding of neural networks and that the long training time will be prohibitive of the students finishing. To address that, maybe an idea could be:
> 
>     1. Shorten the "Build a CNN model" description to just give the bare bones info on the layers of the network and then point them to an appendix with the full details as you have them written.

Thank you for your comment.  I have moved the *Model Intuition* information to an appendix as suggested, and created an in-text hyperlink to the content.

> 
>     2. You are using a very small learning rate in Adam. Is that needed? I tried with a much larger rate (0.05) and the training after 1 epoch gave a test MSE of 0.01. Maybe you could start them with a larger rate and then also include in the optional that they can try with a smaller rate and more epochs if they'd like to get really good results.

Thank you for your comment.  Yes, the small learning rate is required to adequately capture the thermal noise in the data and not have the gradient collapse.  When using a larger learning rate, the MSE quickly stops improving after a few epochs, and plotting the predictions shows that the model has not learned the data.  Learning rates from 1E-2 to 1E-9 were tested with various other hyperparameter combinations over a two-week period in order to determine the easiest initial configuration for the students; the MSE drops to 1E-2 in under 10 minutes for LR=1E-8.   When using the GoogleColab GPU, the total time to train took about 1hr 10 minutes.  Since I warn the students at the beginning of the notebook and again at the beginning of the exercise that it may take several hours, I feel comfortable leaving them with the setup as it is.  I have also left a note to the instructors in the solutions rubric that, depending on the computational resources available, adjustments may need to be made.  Finally, the end of notebook questions focus on asking the students to understand the CNN algorithm instead of optimizing it.
